source/agora/utils/PrettyPrinter.d

```diff
@@ -78,10 +78,65 @@ public auto prettify (T) (const auto ref T input) nothrow
         return ConsensusDataFmt(input);
     else static if (is(T : const QuorumConfig))
         return QuorumConfigFmt(input);
+    else static if (is(T == struct))  // recurse into fields and auto-prettify
+    {
+        struct Formatted
+        {
+            public void toString (scope void delegate (scope const char[]) @safe sink) @safe nothrow
+            {
+                import std.traits;
+                try
+                {
+                    sink("{ ");
+                    foreach (idx, field; FieldNameTuple!T)
+                    {
+                        formattedWrite(sink, field ~ ": %s", prettify(__traits(getMember, input, field)));
```

Comment:
Gonna fix this allocation, hold up.

Reply:
Is it allocating ? I think the compiler const fold. But better to make it explicitly use `%s: %s` yes.

Reply:
You know that's even better, I'll do that.